pkg/ccl/changefeedccl/compression.go

```diff
@@ -17,24 +18,115 @@ import (
 	"github.com/klauspost/pgzip"
 )
 
-var useFastGzip = settings.RegisterBoolSetting(
-	settings.ApplicationLevel,
-	"changefeed.fast_gzip.enabled",
-	"use fast gzip implementation",
-	metamorphic.ConstantWithTestBool(
-		"changefeed.fast_gzip.enabled", true,
-	),
-	settings.WithPublic)
+var (
+	useFastGzip = settings.RegisterBoolSetting(
+		settings.ApplicationLevel,
+		"changefeed.fast_gzip.enabled",
+		"use fast gzip implementation",
+		metamorphic.ConstantWithTestBool(
+			"changefeed.fast_gzip.enabled", true,
+		),
+		settings.WithPublic)
+	gzipEncoderPool     = sync.Pool{}
+	fastGzipEncoderPool = sync.Pool{}
+	zstdEncoderPool     = sync.Pool{}
+	gzipDecoderPool     = sync.Pool{}
+	zstdDecoderPool     = sync.Pool{}
+)
 
 type compressionAlgo string
 
-const sinkCompressionGzip compressionAlgo = "gzip"
-const sinkCompressionZstd compressionAlgo = "zstd"
+const (
+	sinkCompressionGzip compressionAlgo = "gzip"
+	sinkCompressionZstd compressionAlgo = "zstd"
+)
 
 func (a compressionAlgo) enabled() bool {
 	return a != ""
 }
 
+type encoder interface {
+	io.WriteCloser
+	Reset(io.Writer)
+	Flush() error
+}
+
+type decoder interface {
+	io.ReadCloser
+	Reset(io.Reader) error
+}
+
+// encWrapper wraps an encoder with encoders pool.
+type encWrapper struct {
+	encoder encoder
+	pool    *sync.Pool
+}
+
+// Close flushes and closes the underlying encoder, resets it, and returns the wrapper to the pool.
+// This differs from the inner encoder's Close by adding pool management - the wrapper
+// is recycled after closing rather than being discarded.
+func (e *encWrapper) Close() error {
+	var err error
+	if fErr := e.encoder.Flush(); fErr != nil {
+		err = fErr
+	}
+	if cErr := e.encoder.Close(); cErr != nil {
+		if err != nil {
+			return errors.CombineErrors(err, cErr)
+		} else {
+			err = cErr
+		}
+		return errors.Wrap(err, "failed to close writer")
+	}
+	e.encoder.Reset(nil)
+	e.pool.Put(e)
+	return nil
+}
+
+func (e *encWrapper) Write(p []byte) (int, error) {
+	return e.encoder.Write(p)
+}
+
+func (e *encWrapper) Reset(w io.Writer) {
+	e.encoder.Reset(w)
+}
+
+// decWrapper wraps a decoder with decoders pool.
+type decWrapper struct {
+	decoder decoder
+	pool    *sync.Pool
+}
+
+// Close closes the underlying decoder and returns the wrapper to the pool.
+func (d *decWrapper) Close() error {
+	if err := d.decoder.Close(); err != nil {
+		return errors.Wrap(err, "failed to close reader")
+	}
+
+	d.pool.Put(d)
```

Comment:
Do you not want to reset it to nil here too? whether we do that or not, they should have the same behaviour, no?

Reply:
It's a decoder if you reset it with nil and it's gzip or pgzip next time you take it form pool it'll hang or reading data even if you try to reset it with correct writer. That is what I see in the tests.

Reply:
Shall I add the error saying that nil as a writer to reset is not allowed?

Reply:
That's weird, maybe a bug in the library.. Sure, let's document that, thanks.